build.rs

```diff
@@ -0,0 +1,133 @@
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+//! Write out a file of integration tests which tests every pair of example
+//! files in the `tests/` directory.
+
+#![warn(missing_docs)]
+
+extern crate glob;
+#[macro_use]
+extern crate lazy_static;
+
+use std::fs::File;
+use std::io::Write;
+use std::path::Path;
+
+use glob::glob;
+
+/// Directory containing example files for testing and to which the generated
+/// tests should be written.
+const DIR: &str = "tests/";
+
+lazy_static! {
+    /// Glob all example files in the `tests/` directory.
+    /// Filter out files which have been seeded with deliberate lexical or
+    /// syntactic errors.
+    static ref TEST_FILES: Vec<String> = glob(&format!("{}*.calc", DIR))
+        .expect("Failed to read glob pattern")
+        .into_iter()
+        .chain(
+            glob(&format!("{}*.java", DIR))
+                .expect("Failed to read glob pattern")
+                .into_iter(),
+        )
+        .chain(
+            glob(&format!("{}*.txt", DIR))
+                .expect("Failed to read glob pattern")
+                .into_iter(),
+        )
+        .map(|res| res.unwrap().to_str().unwrap().to_string())
+        .filter(|name| !name.contains("_err"))
+        .collect::<Vec<String>>();
+}
+
+/// Generate a file of integration tests which run a diff (with both Null and
+/// Myers matchers) on every pair of test files in the `tests/` directory.
+/// Include diffs of files against themselves. This test takes a long time to
+/// run (well over 5 minutes) and so should not be run on CI infrastructure such
+/// as Travis.
+///
+/// Run with: `cargo test --no-fail-fast  -- --ignored --nocapture`
+fn main() {
+    let mut outfile = File::create(Path::new(&DIR).join("test_generated.rs")).unwrap();
+    write!(
+        outfile,
+        r#"
+extern crate diffract;
+
+use diffract::myers_matcher::MyersConfig;
+use diffract::null_matcher::NullConfig;
+
+mod common;
+use common::check_files;
+
+    "#
+    ).unwrap();
+    let mut duplicates: Vec<&str> = vec![]; // Avoid duplicate unit tests.
```

Comment:
Why would we have duplicate tests? If we find one, shouldn't we error?

Reply:
I think this was left over from a time when the test names were less sophisticated, so files like `Hello.txt` and `Hello.java` would produce four tests called `test_globbed_hello_hello`!